torch/distributed/rpc/__init__.py

```diff
@@ -198,6 +198,25 @@ def init_rpc(
         _init_rpc_backend(backend, store, name, rank, world_size, rpc_backend_options)
 
 
+    @api._require_initialized
+    def barrier(worker_names=None):
+        r"""
+        Synchronizes local and remote RPC processes.
+
+        This will block until all local and remote RPC processes specified under worker_names
+        reach this method to wait for all outstanding work to complete.
+
+        Args:
+            worker_names (List[str], optional): The set of workers to synchronize. If ``None``, the
+            set will be all workers. Default is ``None``.
```

Comment:
> If ``None``, the set will be all workers. Default is ``None``.

This will be a little tricky after we add dynamic membership. As that would require all processes always have a consistent view on the current gang membership, which can be costly and hard. To avoid that potential problem, shall we require `worker_names` to be not None for now?

Reply:
Two Questions: 

1. what if there are concurrent barriers, e.g., two current barriers between [work0, work1] and [worker2, worker3]. Will this still work?
2. If one of the worker specified a wrong list, what gonna happen?

Reply:
Great points. I will change to the worker_names to be a required argument. For the questions:

1. Yes, this works.
2. If the list has workers that doesn't exist, then it would hang and eventually timeout.

Reply:
> If the list has workers that doesn't exist, then it would hang and eventually timeout.

I was also concerned about the following case:

worker0: `worker_names=["worker0", "worker1", "worker2"]`
worker1: `worker_names=["worker0", "worker1", "worker2"]`
worker2: `worker_names=["worker1", "worker2"]`

Do we know what gonna happen in this case? I guess it will also timeout as `worker0` serves as the leader for the first two and `worker1` serves as the leader for `worker2`.

Reply:
Yes, this situation would also time out, which I think should be expected behavior of the API? For the implementation right now, if one worker calls `barrier(<list_of_workers>)` it expects all the workers in <ilst_of_workers> to also call `barrier(<list_of_workers>)` otherwise it will hang.